sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +129,49 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    // As the analyzer makes sure that the list of In is already of the same data type, then the
+    // rule can simply check the first literal in `in.list` can implicitly cast to `toType` or not,
+    // and this rule doesn't convert in when `in.list` is empty.
+    case in @ In(Cast(fromExp, toType: NumericType, _), list @ Seq(firstLit, _*))
+        if canImplicitlyCast(fromExp, toType, firstLit.dataType) && in.inSetConvertible =>
+      val (newValueList, exp) =
+        list.map(lit => unwrapCast(EqualTo(in.value, lit)))
+          .partition {
+            case EqualTo(_, _: Literal) => true
+            case _ => false
```

Comment:
On second thoughts, this is a bit risky if unexpected patterns found. Sorry for bothering you, but how about this instead?
```
            case And(IsNull(_), Literal(null, BooleanType)) => false
            case _ => throw new IllegalStateException("....")
```

Reply:
SGTM, will do.